### DISCUSSION TITLE
Library file for STM32

### DISCUSSION THREAD

**Discussion by DariusBotha95** [ARM-software/CMSIS_5#1214] [Q&A]
Hello,

can someone please help me to locate (or make) the .a library file that I can use in STM32CubeIDE as I did with CMSIS_4?

Thank you for your time and effort.

**Comment by JonatanAntoni** [comment]
Hi @dariusbothaECAM,

May I ask you to be slightly more verbose, please? What library are you actually looking for? CMSIS is a collection of components and we had pre-built libraries for RTX4, RTX5 and DSP in the past. Maintaining these has turned out not to be affortable anymore. Hence we recently removed the CMSIS-DSP libraries and only ship the source code variant for this component.

You can configure the CMSIS-DSP library for your application and compile it into a library archive if required. Is this what you are after?

Cheers,
Jonatan

**Reply by DariusBotha95** [reply]
Hi @JonatanAntoni, I'm sorry. I realise that the question is very vague.

Yes, I want to use the DSP library on both an STM32F4 and F7. I couldn't find how to make the archive ( or if I did, it went completely over my head)

I am on a Windows machine and have Git and 7-zip installed ( I tried following the create software pack route - I couldn't get it to work, and I'm not even sure if that is the correct way to do it.) 

If you don't mind, could you point me in the correct direction or help me with such an archive?

Thank you for your time.

Much appreciated,
Darius

**Reply by JonatanAntoni** [reply]
Hi @dariusbothaECAM,

As I said, we dropped pre-built libraries for CMSIS-DSP from the pack. You need to build DSP from source code yourself. You can either add the source files directly to your project (e.g. by selecting the DSP component in an RTE-enabled IDE) or create a separate library project. In any case, you need to configure the DSP library for your use case. If you have detailed questions about using the DSP library in a specific scenario the DSP team can give you advanced guidance.

Cheers,
Jonatan

**Reply by DariusBotha95** [reply]
Thank you for your reply. I'll try that :)

**Reply by DariusBotha95** [reply]
Hi @JonatanAntoni, I don't seem to come right in building the DSP library from source (I am an absolute novice with CMake and Linux stuff still) Is there a tutorial I can follow in building the library?

Thanks again for your time.

Kind regards,
Darius

**Reply by JonatanAntoni** [reply]
Let us check with @christophe0606 who should know about all the details.

**Reply by christophe0606** [reply]
Hi @dariusbothaECAM , 

you confirm you want to use cmake and not the official CMSIS pack way ?

If you build with cmake, are you trying to build an independent library or are you trying to build CMSIS-DSP as part of a bigger cmake build (the application) ?

**Reply by DariusBotha95** [reply]
Hi @christophe0606, Yes I'm trying to CMake build a DSP library as with CMSIS4 ( an .a file that I include in STMCUBEIDE together with the arm_math.h file) in order to use the DSP functions - I assume it will be an independent library.

Thank you for your assistance.

Kind regards,
Darius

**Reply by christophe0606** [reply]
Hi @dariusbothaECAM,

If you look at DSP/README.md, there is a section "Building only the CMSIS-DSP library" which is explaining how to do it.

You need to create 2 new folders:

- A folder containing a new CMakeLists.txt as described in the README. Let's call this folder CMSISDSPBuilding
- Inside, another folder "build" where the libraries and objects will be created

From build, you type a command cmake to generate the make. Then you can build with a make all and it will generate a lib.a per subfolder of CMSIS-DSP (libBasicMaths.a etc ...)

It is important that your cmake is version 3.14. Some people have reported problems with more recent versions but I never had the time to investigate so far.

The cmake command could be something like that:

```
cmake -DCMAKE_PREFIX_PATH="Path to folder containing the compiler bin folder" 
  -DCMAKE_TOOLCHAIN_FILE="Absolute path to armac6.cmake inside CMSIS/DSP" 
  -DROOT="Path to CMSIS folder (the one containing Device and CMSIS)" 
  -DARM_CPU="cortex-m4" 
  -DLOOPUNROLL=ON 
  -DCMAKE_BUILD_TYPE=Release
  -DCMAKE_C_FLAGS_RELEASE="-std=c11 -Ofast -ffast-math -DNDEBUG -Wall -Wextra" \
  -DCMAKE_CXX_FLAGS_RELEASE="-fno-rtti -std=c++11 -Ofast -ffast-math -DNDEBUG -Wall -Wextra -Wno-unused-parameter" 
  -G "Unix Makefiles" ..
```

If you're building with gcc, replace armac6.cmake with gcc.cmake

The cmake command must be on one line. I have written it on multiple lines so that it is easier to read here.

On Linux, you can use \ at end of line if you want to write it as a multiline command.

**Reply by JonatanAntoni** [reply]
@DariusBotha95, @christophe0606,

It the latest reply the answer for the original question?
If yes, it might make sense to put it into an individual anwer and mark is as "the answer".

Thanks,
Jonatan

**Reply by DariusBotha95** [reply]
> Hi @dariusbothaECAM,
> 
> If you look at DSP/README.md, there is a section "Building only the CMSIS-DSP library" which is explaining how to do it.
> 
> You need to create 2 new folders:
> 
> * A folder containing a new CMakeLists.txt as described in the README. Let's call this folder CMSISDSPBuilding
> * Inside, another folder "build" where the libraries and objects will be created
> 
> From build, you type a command cmake to generate the make. Then you can build with a make all and it will generate a lib.a per subfolder of CMSIS-DSP (libBasicMaths.a etc ...)
> 
> It is important that your cmake is version 3.14. Some people have reported problems with more recent versions but I never had the time to investigate so far.
> 
> The cmake command could be something like that:
> 
> ```
> cmake -DCMAKE_PREFIX_PATH="Path to folder containing the compiler bin folder" 
>   -DCMAKE_TOOLCHAIN_FILE="Absolute path to armac6.cmake inside CMSIS/DSP" 
>   -DROOT="Path to CMSIS folder (the one containing Device and CMSIS)" 
>   -DARM_CPU="cortex-m4" 
>   -DLOOPUNROLL=ON 
>   -DCMAKE_BUILD_TYPE=Release
>   -DCMAKE_C_FLAGS_RELEASE="-std=c11 -Ofast -ffast-math -DNDEBUG -Wall -Wextra" \
>   -DCMAKE_CXX_FLAGS_RELEASE="-fno-rtti -std=c++11 -Ofast -ffast-math -DNDEBUG -Wall -Wextra -Wno-unused-parameter" 
>   -G "Unix Makefiles" ..
> ```
> 
> If you're building with gcc, replace armac6.cmake with gcc.cmake
> 
> The cmake command must be on one line. I have written it on multiple lines so that it is easier to read here.
> 
> On Linux, you can use \ at end of line if you want to write it as a multiline command.

Thank you. This helped me!
